hyperspace/cosmos/src/provider.rs

```diff
@@ -0,0 +1,1127 @@
+use ibc_rpc::PacketInfo;
+use ics07_tendermint::{
+	client_message::ClientMessage, client_state::ClientState, consensus_state::ConsensusState,
+};
+use pallet_ibc::light_clients::{
+	AnyClientMessage, AnyClientState, AnyConsensusState, HostFunctionsManager,
+};
+use primitives::{mock::LocalClientTypes, Chain, IbcProvider, UpdateType};
+use prost::Message;
+use std::{pin::Pin, str::FromStr, time::Duration};
+use tendermint::block::Height as TmHeight;
+pub use tendermint::Hash;
+use tendermint_rpc::{
+	endpoint::tx::Response,
+	event::{Event, EventData},
+	query::{EventType, Query},
+	Client, Error as RpcError, Order, SubscriptionClient, WebSocketClient,
+};
+use tokio::time::sleep;
+
+#[derive(Clone, Debug)]
+pub enum FinalityEvent {
+	Tendermint { from: TmHeight, to: TmHeight },
+}
+
+#[derive(Clone, Debug)]
+pub struct TransactionId<Hash> {
+	pub hash: Hash,
+}
+
+#[async_trait::async_trait]
+impl<H> IbcProvider for CosmosClient<H>
+where
+	H: Clone + Send + Sync + 'static,
+{
+	type FinalityEvent = FinalityEvent;
+	type TransactionId = TransactionId<Hash>;
+	type AssetId = String;
+	type Error = Error;
+
+	async fn query_latest_ibc_events<C>(
+		&mut self,
+		finality_event: Self::FinalityEvent,
+		counterparty: &C,
+	) -> Result<(Vec<Any>, Vec<IbcEvent>, UpdateType), anyhow::Error>
+	where
+		C: Chain,
+	{
+		let finality_event_height = match finality_event {
+			FinalityEvent::Tendermint { from: _, to } => to,
+		};
+		let client_id = self.client_id();
+		let latest_cp_height = counterparty.latest_height_and_timestamp().await?.0;
+		let latest_cp_client_state =
+			counterparty.query_client_state(latest_cp_height, client_id.clone()).await?;
+		let client_state_response = latest_cp_client_state
+			.client_state
+			.ok_or_else(|| Error::Custom("counterparty returned empty client state".to_string()))?;
+		let client_state =
+			ClientState::<HostFunctionsManager>::decode_vec(&client_state_response.value)
+				.map_err(|_| Error::Custom("failed to decode client state response".to_string()))?;
+		let latest_cp_client_height = client_state.latest_height().revision_height;
+		let latest_height = self.latest_height_and_timestamp().await?.0;
+		let latest_revision = latest_height.revision_number;
+
+		let mut ibc_events: Vec<IbcEvent> = vec![];
+		let from = TmHeight::try_from(latest_cp_client_height).unwrap();
+		let to = finality_event_height;
+		log::info!(target: "hyperspace_cosmos", "Getting blocks {}..{}", from, to);
+
+		// query (exclusively) up to `to`, because the proof for the event at `to - 1` will be
+		// contained at `to` and will be fetched below by `msg_update_client_header`
+		for height in from.value()..to.value() {
+			let block_results =
+				self.rpc_client.block_results(TmHeight::try_from(height)?).await.map_err(|e| {
+					Error::from(format!(
+						"Failed to query block result for height {:?}: {:?}",
+						height, e
+					))
+				})?;
+			// TODO: handle block_results.begin_block_events, block_results.end_block_events?
```

Comment:
what does this TODO mean?

Reply:
So that's the part I'm not sure about, I'll check how it's done in `hemes`

Reply:
Done